source/comp/markv_codec.cpp

```diff
@@ -0,0 +1,1194 @@
+    case SPV_OPERAND_TYPE_OPTIONAL_LITERAL_INTEGER:
+      return 6;
+    case SPV_OPERAND_TYPE_CAPABILITY:
+      return 6;
+    case SPV_OPERAND_TYPE_SOURCE_LANGUAGE:
+    case SPV_OPERAND_TYPE_EXECUTION_MODEL:
+      return 3;
+    case SPV_OPERAND_TYPE_ADDRESSING_MODEL:
+    case SPV_OPERAND_TYPE_MEMORY_MODEL:
+      return 2;
+    case SPV_OPERAND_TYPE_EXECUTION_MODE:
+      return 6;
+    case SPV_OPERAND_TYPE_STORAGE_CLASS:
+      return 4;
+    case SPV_OPERAND_TYPE_DIMENSIONALITY:
+    case SPV_OPERAND_TYPE_SAMPLER_ADDRESSING_MODE:
+      return 3;
+    case SPV_OPERAND_TYPE_SAMPLER_FILTER_MODE:
+      return 2;
+    case SPV_OPERAND_TYPE_SAMPLER_IMAGE_FORMAT:
+      return 6;
+    case SPV_OPERAND_TYPE_FP_ROUNDING_MODE:
+    case SPV_OPERAND_TYPE_LINKAGE_TYPE:
+    case SPV_OPERAND_TYPE_ACCESS_QUALIFIER:
+    case SPV_OPERAND_TYPE_OPTIONAL_ACCESS_QUALIFIER:
+      return 2;
+    case SPV_OPERAND_TYPE_FUNCTION_PARAMETER_ATTRIBUTE:
+      return 3;
+    case SPV_OPERAND_TYPE_DECORATION:
+    case SPV_OPERAND_TYPE_BUILT_IN:
+      return 6;
+    case SPV_OPERAND_TYPE_GROUP_OPERATION:
+    case SPV_OPERAND_TYPE_KERNEL_ENQ_FLAGS:
+    case SPV_OPERAND_TYPE_KERNEL_PROFILING_INFO:
+      return 2;
+    case SPV_OPERAND_TYPE_FP_FAST_MATH_MODE:
+    case SPV_OPERAND_TYPE_FUNCTION_CONTROL:
+    case SPV_OPERAND_TYPE_LOOP_CONTROL:
+    case SPV_OPERAND_TYPE_IMAGE:
+    case SPV_OPERAND_TYPE_OPTIONAL_IMAGE:
+    case SPV_OPERAND_TYPE_OPTIONAL_MEMORY_ACCESS:
+    case SPV_OPERAND_TYPE_SELECTION_CONTROL:
+      return 4;
+    default:
+      return 0;
+  }
+  return 0;
+}
+
+// Defines and returns current MARK-V version.
+uint32_t GetMarkvVersion() {
+  const uint32_t kVersionMajor = 1;
+  const uint32_t kVersionMinor = 0;
+  return kVersionMinor | (kVersionMajor << 16);
+}
+
+class CommentLogger {
+ public:
+  void AppendText(const std::string& str) {
+    Append(str);
+    delimiter_ = false;
+  }
+
+  void AppendTextNewLine(const std::string& str) {
+    Append(str);
+    Append("\n");
+    delimiter_ = false;
+  }
+
+  void AppendBitSequence(const std::string& str) {
+    if (delimiter_)
+      Append("-");
+    Append(str);
+    delimiter_ = true;
+  }
+
+  void AppendWhitespaces(size_t num) {
+    Append(std::string(num, ' '));
+    delimiter_ = false;
+  }
+
+  void NewLine() {
+    Append("\n");
+    delimiter_ = false;
+  }
+
+  std::string GetText() const {
+    return ss_.str();
+  }
+
+ private:
+  void Append(const std::string& str) {
+    ss_ << str;
+    // std::cerr << str;
+  }
+
+  std::stringstream ss_;
+  bool delimiter_ = false;
+};
+
+spv_text CreateSpvText(const std::string& str) {
+  spv_text out = new spv_text_t();
+  assert(out);
+  char* cstr = new char[str.length() + 1];
+  assert(cstr);
+  std::strncpy(cstr, str.c_str(), str.length());
+  cstr[str.length()] = '\0';
+  out->str = cstr;
+  out->length = str.length();
+  return out;
+}
+
+// Base class for MARK-V encoder and decoder.
```

Comment:
Since this is your contact point with the valdiation state, should say how you'll be using it.  It seems you're using it to maintain certain bookkeeping (and give examples: id bound, unresolved forward ids, vector of parsed instructions...), and to help process diagnostics.